Project_1_EKF/src/tools.cpp

```diff
@@ -19,7 +19,28 @@ VectorXd Tools::CalculateRMSE(const vector<VectorXd> &estimations,
 
 MatrixXd Tools::CalculateJacobian(const VectorXd& x_state) {
   /**
-  TODO:
     * Calculate a Jacobian here.
+    * Assume x_state has 4 states (2d position and 2d velocity)
   */
+  MatrixXd H_(3, 4);
+
+  double px = x_state(0);
+  double py = x_state(1);
+  double vx = x_state(2);
+  double vy = x_state(3);
+
+  double r2 = px * px + py * py;
+  double r = sqrt(r2);
+
+  // check division by zero
+  if (fabs(r2) < 0.0001) {
+    cout << "Calculation Jacobian Error: division by Zero" << endl;
+    return H_;
+  }
+
+  H_ << px / r, py / r, 0, 0,
+          - py / r, px / r, 0, 0,
```

Comment:
The denominator for the second row elements should be `r2` rather than `r`?

Reply:
good catch, thanks